File: templates/problems/01_arrays_hashing/05_top_k_frequent/python/solution.py
def solve_top_k_frequent(nums: list[int], k: int) -> list[int]:
    """Find k most frequent elements using bucket sort.

    Time: O(n), Space: O(n)
    """
    freq: dict[int, int] = {}
    for n in nums:
        freq[n] = freq.get(n, 0) + 1

    # Bucket sort: index = frequency, value = list of numbers with that frequency
    buckets: list[list[int]] = [[] for _ in range(len(nums) + 1)]
    for num, count in freq.items():
        buckets[count].append(num)

    result: list[int] = []
    for i in range(len(buckets) - 1, -1, -1):
        if len(result) >= k:
            break
        result.extend(buckets[i])
    return result[:k]

File: templates/problems/01_arrays_hashing/05_top_k_frequent/typescript/solution.ts
/**
 * Find k most frequent elements using bucket sort.
 * Time: O(n), Space: O(n)
 */
export function solveTopKFrequent(nums: number[], k: number): number[] {
    const freq = new Map<number, number>();
    for (const n of nums) {
        freq.set(n, (freq.get(n) ?? 0) + 1);
    }

    // Bucket sort: index = frequency, value = list of numbers with that frequency
    const buckets: number[][] = Array.from({ length: nums.length + 1 }, () => []);
    for (const [num, count] of freq) {
        buckets[count].push(num);
    }

    const result: number[] = [];
    for (let i = buckets.length - 1; i >= 0 && result.length < k; i--) {
        result.push(...buckets[i]);
    }
    return result.slice(0, k);
}
